ipa-core/src/protocol/prss/crypto.rs

```diff
@@ -8,12 +8,67 @@ use sha2::Sha256;
 use x25519_dalek::{EphemeralSecret, PublicKey};
 
 use crate::{
-    ff::{Field, GaloisField},
-    secret_sharing::replicated::{
-        semi_honest::AdditiveShare as Replicated, ReplicatedSecretSharing,
+    ff::Field,
+    secret_sharing::{
+        replicated::{semi_honest::AdditiveShare as Replicated, ReplicatedSecretSharing},
+        SharedValue,
     },
 };
 
+/// Trait for random generation from random u128s.
+///
+/// It was previously assumed that our fields were of order << 2^128, in which case
+/// `Field::truncate_from` can be used for this purpose. This trait makes the contract explicit.
+pub trait FromRandomU128 {
+    /// Generate a random value of `Self` from a uniformly-distributed random u128.
+    fn from_random_u128(src: u128) -> Self;
+}
+
+/// Trait for things that can be generated by PRSS.
+///
+/// The exact semantics of the generation depend on the value being generated.
+///
+/// At a high level, there are two kinds of PRSS generation:
+///  1. Raw values: In this case, two values are generated, one using the randomness that is shared
+///     with the left helper, and one with the randomness that is shared with the right helper.
+///     Thus, one of the generated values is known to both us and the left helper, and likewise for
+///     the right helper.
+///  2. Secret sharings: In this case, a single secret-shared random value is generated. The value
+///     returned by `FromPrss` is our share of that sharing. Within `FromPrss`, the randomness shared
+///     with the left and right helpers is used to construct the sharing.
+///
+/// In the first case, `FromPrss` is implemented for a tuple type, while in the second case,
+/// `FromPrss` is implemented for a secret-shared type.
+pub trait FromPrss: Sized {
```

Comment:
I don't know how, but it would be great if we could capture the fact that values need to be sampled uniformly. Right now, someone could just naively sample a single byte from PRSS and then expand it to any `T` by padding, masking, etc.

Feels like `Distribution<Standard>` at least puts a warning sign, but I don't know if we could put it here. Maybe a rustdoc comment for `FromPrss`?

Reply:
Yeah. I originally had something in the comment like "don't implement this if it could be a footgun", but took it out because it seemed unhelpful. It's a bit wordy, but here's what I have now:

```
/// The exact semantics of the generation depend on the value being generated, but like
/// `rand::distributions::Standard`, a uniform distribution is typical. When implementing
/// this trait, consider the consequences if the implementation were to be used in
/// an unexpected way. For example, an implementation that draws from a subset of the
/// possible values could be dangerous, if used in an unexpected context where
/// security relies on sampling from the full space.
```

Reply:
I was also thinking of requiring Distribution<Standard> being implemented for PRSS and then generating a seed using AES and using the seed to define an rng which then generates the desired random value. It seems that you only solve the problem for BA256, but we need PRSS also for small values like BA8 and so on. I would prefer to put it in the macro and solve the issue in general, for any BA type.

Reply:
I feel like another level of AES can only slow things down, but maybe I'm not understanding.

I'm still looking at this stuff in the context of PRSS generation of an array. There are a bunch of issues to figure out. I am exploring the path of supporting random generation from an integer multiple of u128s. (Rationale being that we can always vectorize to get things that wide.) But there are a bunch of issues (trait bounds, limited const generic support) that I haven't figured out yet. Generating from a multiple of u128s at least keeps the PRSS indices reasonably well-defined. If we want to generate below the level of u128s, we may have to revise what a PRSS index means.

> we need PRSS also for small values like BA8 and so on

We have it for the small BAs currently via `FromRandomU128` and `Field::truncate_from`. When we make them not Fields, we can move the contents of `Field::truncate_from` into `FromRandomU128`. The generation is not efficient in terms of PRSS bits consumed, but neither is generation of Fp32BitPrimes.

Reply:
To offer something more concrete, this is what I'm working with right now:

```
pub trait FromRandom {
    type Source; // == [u128; N]
    fn len() -> usize; // returns N

    /// Generate a random value of `Self` from a uniformly-distributed `Self::Source`.
    fn from_random(src: Self::Source) -> Self;
}
```

This seems like it can work, but is a bit unwieldy. If we wanted to use `[u8; N]` instead of `[u128; N]`, we could.

Reply:
I think that would make sense. I am a bit worried about adding more traits around PRSS. Ideally generate value would give you the N random values.

Reply:
I opened #884 to continue this discussion.